openssl/src/hash.rs

```diff
@@ -192,34 +202,95 @@ impl Drop for Hasher {
     fn drop(&mut self) {
         unsafe {
             if self.state != Finalized {
-                drop(self.finish());
+                drop(self.finish2());
             }
             EVP_MD_CTX_free(self.ctx);
         }
     }
 }
 
-/// Computes the hash of the `data` with the hash `t`.
+/// The resulting bytes of a digest.
+///
+/// This type derefs to a byte slice - it exists to avoid allocating memory to
+/// store the digest data.
+#[derive(Copy)]
+pub struct DigestBytes {
+    buf: [u8; ffi::EVP_MAX_MD_SIZE as usize],
+    len: usize,
+}
+
+impl Clone for DigestBytes {
+    #[inline]
+    fn clone(&self) -> DigestBytes {
+        *self
+    }
+}
+
+impl Deref for DigestBytes {
+    type Target = [u8];
+
+    #[inline]
+    fn deref(&self) -> &[u8] {
+        &self.buf[..self.len]
+    }
+}
+
+impl DerefMut for DigestBytes {
+    #[inline]
+    fn deref_mut(&mut self) -> &mut [u8] {
+        &mut self.buf[..self.len]
+    }
+}
+
+impl AsRef<[u8]> for DigestBytes {
+    #[inline]
+    fn as_ref(&self) -> &[u8] {
+        self.deref()
+    }
+}
+
+impl PartialEq for DigestBytes {
+    #[inline]
+    fn eq(&self, other: &DigestBytes) -> bool {
+        &**self == &**other
```

Comment:
Ah sure, good call. I'm not sure how useful this impl is anyway since it seems a bit rare to be comparing two `DigestBytes` rather than e.g. a `DigestBytes` and a `Vec<u8>`. I only added it for testing convenience but that's easy to fix. Maybe it's best to just remove it?